src/main/java/org/mule/extension/http/internal/request/HttpPollingSource.java

```diff
@@ -181,53 +179,55 @@ private String getResolvedUri() {
+          .ifPresent(idExp -> pollItem.setId(getItemId(fullResponse, idExp, watermark, item, expressionLanguage)));
+      expressions.getWatermarkExpression()
+          .ifPresent(wExp -> pollItem.setWatermark(getItemWatermark(fullResponse, wExp, watermark, item, expressionLanguage)));
     };
   }
 
-  private void sendRequest(PollContext<String, HttpResponseAttributes> pollContext) {
-    CompletionCallback<InputStream, HttpResponseAttributes> callback =
-        new CompletionCallback<InputStream, HttpResponseAttributes>() {
-
-          @Override
-          public void success(Result<InputStream, HttpResponseAttributes> result) {
-            HttpResponseAttributes attributes = result.getAttributes().orElse(null);
-            MediaType mediaType = result.getMediaType().orElse(ANY);
-            Reference<Boolean> atLeastOneResult = new Reference<>(false);
-            getItems(result.getOutput(), mediaType, attributes).forEach(item -> {
-              atLeastOneResult.set(true);
-              pollContext.accept(getPollingItemConsumer(item));
-            });
-
-            if (!atLeastOneResult.get()) {
-              LOGGER.debug("Empty result in HTTP Polling Source at {} of uri {}", location.getRootContainerName(), resolvedUri);
-            }
-          }
-
-          @Override
-          public void error(Throwable throwable) {
-            LOGGER.error("There was an error in HTTP Polling Source at {} of uri '{}'", location.getRootContainerName(),
-                         resolvedUri, throwable);
-          }
-        };
+  private void pollResult(PollContext<String, HttpResponseAttributes> pollContext,
+                          Result<InputStream, HttpResponseAttributes> result, Serializable currentWatermark) {
+    HttpResponseAttributes attributes = result.getAttributes().orElse(null);
+    MediaType mediaType = result.getMediaType().orElse(ANY);
+    Charset charset = mediaType.getCharset().orElse(defaultCharset());
+    TypedValue<String> response = toTypedValue(IOUtils.toString(result.getOutput(), charset), mediaType, charset);
+    LOGGER.debug("Received response at {}: {} and headers {}", location.getRootContainerName(), response,
+                 attributes.getHeaders());
+
+    Reference<Boolean> atLeastOneResult = new Reference<>(false);
+    getItems(response, attributes, currentWatermark, expressions.getSplitExpression(), expressionLanguage).forEach(item -> {
+      atLeastOneResult.set(true);
+      pollContext.accept(getPollingItemConsumer(response, item, currentWatermark));
+    });
+
+    if (!atLeastOneResult.get()) {
+      LOGGER.debug("Empty result in HTTP Polling Source at {} of uri {}", location.getRootContainerName(), resolvedUri);
+    }
+  }
+
+  private void sendRequest(PollContext<String, HttpResponseAttributes> pollContext) throws InterruptedException {
+    Serializable currentWatermark = pollContext.getWatermark().orElse(null);
+    requestBuilder.updateWatermark(currentWatermark);
```

Comment:
i would assume the `requestBuilder` is more of a template... if bothers me that the watermark is being set into it for every request...

Reply:
But the watermark has to be set for every request because it will change